headless attach + connect results in SIGSEGV on detach

1. What version of Delve are you using (`dlv version`)?
Version: 1.5.1
I actually built at the head revision.

2. What version of Go are you using? (`go version`)?
go version go1.14.3 darwin/amd64

3. What operating system and processor architecture are you using?
MacOS Catalina 10.15.7
1.6 GHz Dual-Core Intel Core i5

4. What did you do?
I attached to a running process and then detached without killing it.

5. What did you expect to see?
Clean shutdown.

6. What did you see instead?
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x438623d]

**Details**

I found this issue when using "attach"+"local" mode in vscode-go.
I was able to reproduce with `dlv attach --headless` and `dlv connect` using `_fixtures/loopprog.go`
```
$ pgrep loopprog
23540
polina-macbookair2:delve polina$ ./dlv attach 23540 --log --log-output=debugger,rpc --headless
2020-12-10T22:23:01-08:00 info layer=rpc Using API v1
API server listening at: 127.0.0.1:63029
2020-12-10T22:23:01-08:00 debug layer=rpc API server pid = 23589
2020-12-10T22:23:01-08:00 info layer=debugger attaching to pid 23540
......
2020-12-10T22:23:47-08:00 debug layer=rpc (async 7) <- RPCServer.State(rpc2.StateIn{"NonBlocking":false})
2020-12-10T22:23:47-08:00 debug layer=rpc (async 7) -> rpc2.StateOut{"State":{"Running":false,"Recording":false,"currentThread":{"id":2102284,"pc":140734987069570,"file":"","line":0,"goroutineID":0,"ReturnValues":null,"CallReturn":false},"Threads":[{"id":2102285,"pc":140734987069570,"file":"","line":0,"goroutineID":0,"ReturnValues":null,"CallReturn":false},{"id":2102286,"pc":140734987069570,"file":"","line":0,"goroutineID":0,"ReturnValues":null,"CallReturn":false},{"id":2102282,"pc":17573529,"file":"/Users/polina/delve/_fixtures/loopprog.go","line":9,"function":{"name":"main.loop","value":17573424,"type":0,"goType":0,"optimized":false},"goroutineID":1,"ReturnValues":null,"CallReturn":false},{"id":2102283,"pc":140734987069270,"file":"","line":0,"goroutineID":0,"ReturnValues":null,"CallReturn":false},{"id":2102284,"pc":140734987069570,"file":"","line":0,"goroutineID":0,"ReturnValues":null,"CallReturn":false}],"NextInProgress":false,"exited":false,"exitStatus":0,"When":""}} error: ""
2020-12-10T22:23:47-08:00 debug layer=rpc <- RPCServer.IsMulticlient(rpc2.IsMulticlientIn{})
2020-12-10T22:23:47-08:00 debug layer=rpc -> *rpc2.IsMulticlientOut{"IsMulticlient":false} error: ""
2020-12-10T22:23:47-08:00 debug layer=rpc <- RPCServer.AttachedToExistingProcess(rpc2.AttachedToExistingProcessIn{})
2020-12-10T22:23:47-08:00 debug layer=rpc -> *rpc2.AttachedToExistingProcessOut{"Answer":true} error: ""
2020-12-10T22:23:49-08:00 debug layer=rpc <- RPCServer.Detach(rpc2.DetachIn{"Kill":false})
2020-12-10T22:23:49-08:00 warning layer=debugger could not set asyncpreemptoff <nil>
2020-12-10T22:23:49-08:00 debug layer=rpc -> *rpc2.DetachOut{} error: ""
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x438623d]

goroutine 1 [running]:
github.com/go-delve/delve/pkg/proc/gdbserial.(*gdbConn).send(0xc0000e2a08, 0xc00011a900, 0xe, 0x83e, 0x400, 0xb)
	/Users/polina/delve/pkg/proc/gdbserial/gdbserver_conn.go:1057 +0x20d
github.com/go-delve/delve/pkg/proc/gdbserial.(*gdbConn).exec(0xc0000e2a08, 0xc00011a900, 0xb, 0x83e, 0x47bb2ae, 0xb, 0x2, 0xb, 0x0, 0x0, ...)
	/Users/polina/delve/pkg/proc/gdbserial/gdbserver_conn.go:1030 +0x53
github.com/go-delve/delve/pkg/proc/gdbserial.(*gdbConn).readMemory(0xc0000e2a08, 0xc0005a8cc8, 0x0, 0x4, 0x11c4e64, 0x479f201, 0xc0005a8cc8)
	/Users/polina/delve/pkg/proc/gdbserial/gdbserver_conn.go:898 +0x36d
github.com/go-delve/delve/pkg/proc/gdbserial.(*gdbProcess).ReadMemory(0xc0000e2a00, 0xc0005a8cc8, 0x4, 0x4, 0x11c4e64, 0xc00053b0e0, 0x4, 0x4073292)
	/Users/polina/delve/pkg/proc/gdbserial/gdbserver.go:1308 +0x5d
github.com/go-delve/delve/pkg/proc.readIntRaw(0x48bb380, 0xc0000e2a00, 0x11c4e64, 0x4, 0x3, 0xc00015f4c8, 0x40e3b81)
	/Users/polina/delve/pkg/proc/variables.go:1607 +0x7c
github.com/go-delve/delve/pkg/proc.(*Variable).loadValueInternal(0xc000012300, 0x0, 0x1, 0x1, 0x40, 0x40, 0xffffffffffffffff, 0x0)
	/Users/polina/delve/pkg/proc/variables.go:1278 +0xa81
github.com/go-delve/delve/pkg/proc.(*Variable).loadValue(0xc000012300, 0x1, 0x1, 0x40, 0x40, 0xffffffffffffffff, 0x0)
	/Users/polina/delve/pkg/proc/variables.go:1191 +0x52
github.com/go-delve/delve/pkg/proc.setAsyncPreemptOff(0xc0000cd9e0, 0x1)
	/Users/polina/delve/pkg/proc/target.go:292 +0x39e
github.com/go-delve/delve/pkg/proc.(*Target).Detach(0xc0000cd9e0, 0xc000541b00, 0x48bd8c0, 0x48a1c90)
	/Users/polina/delve/pkg/proc/target.go:258 +0x17c
github.com/go-delve/delve/service/debugger.(*Debugger).detach(...)
	/Users/polina/delve/service/debugger/debugger.go:418
github.com/go-delve/delve/service/debugger.(*Debugger).Detach(0xc0000aefc0, 0x0, 0x0, 0x0)
	/Users/polina/delve/service/debugger/debugger.go:411 +0x8b
github.com/go-delve/delve/service/rpccommon.(*ServerImpl).Stop(0xc0000aef00, 0x0, 0x0)
	/Users/polina/delve/service/rpccommon/server.go:94 +0x4c
github.com/go-delve/delve/cmd/dlv/cmds.execute(0x5bf4, 0xc0000d9bd0, 0x0, 0x3, 0xc000171260, 0x0, 0x0, 0x3, 0xc0000d9bc0, 0x1, ...)
	/Users/polina/delve/cmd/dlv/cmds/commands.go:880 +0x9ee
github.com/go-delve/delve/cmd/dlv/cmds.attachCmd(0xc0000d1d40, 0xc0000d9bc0, 0x1, 0x4)
	/Users/polina/delve/cmd/dlv/cmds/commands.go:663 +0xfb
github.com/spf13/cobra.(*Command).execute(0xc0000d1d40, 0xc0000d9b40, 0x4, 0x4, 0xc0000d1d40, 0xc0000d9b40)
	/Users/polina/go/pkg/mod/github.com/spf13/cobra@v0.0.0-20170417170307-b6cb39589372/command.go:647 +0x231
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000d1b00, 0xc0000d1b00, 0x47b4941, 0x6)
	/Users/polina/go/pkg/mod/github.com/spf13/cobra@v0.0.0-20170417170307-b6cb39589372/command.go:733 +0x2bc
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/polina/go/pkg/mod/github.com/spf13/cobra@v0.0.0-20170417170307-b6cb39589372/command.go:692
main.main()
	/Users/polina/delve/cmd/dlv/main.go:24 +0xc6
```
```
$ ./dlv connect 127.0.0.1:63029
Type 'help' for list of commands.
(dlv) exit
Would you like to kill the process? [Y/n] n
```

--- a/service/debugger/debugger.go
+++ b/service/debugger/debugger.go
@@ -407,7 +407,9 @@ func (d *Debugger) FunctionReturnLocations(fnName string) ([]uint64, error) {
 func (d *Debugger) Detach(kill bool) error {
 	d.targetMutex.Lock()
 	defer d.targetMutex.Unlock()
-
+	if ok, _ := d.target.Valid(); !ok {
+		return nil
+	}
 	return d.detach(kill)
 }
 

--- a/service/test/integration2_test.go
+++ b/service/test/integration2_test.go
@@ -9,6 +9,7 @@ import (
 	"net/rpc"
 	"net/rpc/jsonrpc"
 	"os"
+	"os/exec"
 	"path/filepath"
 	"reflect"
 	"runtime"
@@ -2110,4 +2111,56 @@ func TestIssue2162(t *testing.T) {
 		_, err = c.Step()
 		assertNoError(err, t, "Step()")
 	})
+}
+
+func TestDetachLeaveRunning(t *testing.T) {
+	// See https://github.com/go-delve/delve/issues/2259
+	if testBackend == "rr" {
+		return
+	}
+
+	listener, clientConn := service.ListenerPipe()
+	defer listener.Close()
+	var buildFlags protest.BuildFlags
+	if buildMode == "pie" {
+		buildFlags |= protest.BuildModePIE
+	}
+	fixture := protest.BuildFixture("testnextnethttp", buildFlags)
+
+	cmd := exec.Command(fixture.Path)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	assertNoError(cmd.Start(), t, "starting fixture")
+	defer cmd.Process.Kill()
+
+	// wait for testnextnethttp to start listening
+	t0 := time.Now()
+	for {
+		conn, err := net.Dial("tcp", "127.0.0.1:9191")
+		if err == nil {
+			conn.Close()
+			break
+		}
+		time.Sleep(50 * time.Millisecond)
+		if time.Since(t0) > 10*time.Second {
+			t.Fatal("fixture did not start")
+		}
+	}
+
+	server := rpccommon.NewServer(&service.Config{
+		Listener:   listener,
+		APIVersion: 2,
+		Debugger: debugger.Config{
+			AttachPid:  cmd.Process.Pid,
+			WorkingDir: ".",
+			Backend:    testBackend,
+		},
+	})
+	if err := server.Run(); err != nil {
+		t.Fatal(err)
+	}
+
+	client := rpc2.NewClientFromConn(clientConn)
+	defer server.Stop()
+	assertNoError(client.Detach(false), t, "Detach")
 }

